go/vt/vtgate/engine/distinct.go

```diff
@@ -105,32 +113,42 @@ func (pt *probeTable) hashCodeForRow(inputRow row) (evalengine.HashCode, error)
 	// > So you want an odd MULT that has plenty of one-bits.
 
 	code := evalengine.HashCode(17)
-	for idx, value := range inputRow {
-		// We use unknown collations when we do not have collation information
-		// This is safe for types which do not require collation information like
-		// numeric types. It will fail at runtime for text types.
-		collation := collations.Unknown
-		if len(pt.colCollations) > idx {
-			collation = pt.colCollations[idx]
+	for i, checkCol := range pt.checkCols {
+		if i >= len(inputRow) {
+			return 0, vterrors.New(vtrpcpb.Code_INTERNAL, "distinct check colls is larger than its input row")
 		}
-		hashcode, err := evalengine.NullsafeHashcode(value, collation, value.Type())
+		col := inputRow[checkCol.Col]
+		hashcode, err := evalengine.NullsafeHashcode(col, checkCol.Collation, col.Type())
 		if err != nil {
-			return 0, err
+			if err != evalengine.UnsupportedCollationHashError || checkCol.WsCol == nil {
+				return 0, err
+			}
+			checkCol = checkCol.SwitchToWeighString()
+			pt.checkCols[i] = checkCol
+			hashcode, err = evalengine.NullsafeHashcode(inputRow[checkCol.Col], checkCol.Collation, col.Type())
+			if err != nil {
+				return 0, err
+			}
 		}
 		code = code*31 + hashcode
 	}
 	return code, nil
 }
 
-func equal(a, b []sqltypes.Value, colCollations []collations.ID) (bool, error) {
-	for i, aVal := range a {
-		collation := collations.Unknown
-		if len(colCollations) > i {
-			collation = colCollations[i]
-		}
-		cmp, err := evalengine.NullsafeCompare(aVal, b[i], collation)
+func (pt *probeTable) equal(a, b sqltypes.Row) (bool, error) {
+	for i, checkCol := range pt.checkCols {
+		cmp, err := evalengine.NullsafeCompare(a[i], b[i], checkCol.Collation)
 		if err != nil {
-			return false, err
+			_, isComparisonErr := err.(evalengine.UnsupportedComparisonError)
+			if !isComparisonErr || checkCol.WsCol == nil {
+				return false, err
+			}
+			checkCol = checkCol.SwitchToWeighString()
+			pt.checkCols[i] = checkCol
```

Comment:
because CheckCols are by-value and not references, this does not update the original plan. maybe we should change it so that the plan remembers that we need to use the weight_string

Reply:
not necessary, as someone can change the type from varchar to int and changing the plan will make the query to not work. so, we should avoid modifying the plan

Reply:
only local execution should remember it.

Reply:
You are correct. I wrote a short unit test, and the plan seems to remember the new checkcol.